src/main/scala/com/fulcrumgenomics/bam/AnnotateUmis.scala

```diff
@@ -0,0 +1,92 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.util.ProgressLogger
+import htsjdk.samtools.{SAMFileWriterFactory, SamReaderFactory}
+
+import scala.collection.JavaConversions._
+
+import com.fulcrumgenomics.cmdline.{FgBioTool, ClpGroups}
+import dagr.commons.CommonsDef.{PathToFastq, PathToBam}
+import dagr.commons.io.Io
+import dagr.commons.util.LazyLogging
+import dagr.sopt._
+import htsjdk.samtools.fastq.FastqReader
+
+@clp(description =
+  """
+    |Annotated existing BAM files with UMIs (Unique Molecular Indices, aka Molecular IDs, Molecular barcodes). Takes
+    |an existing BAM and a FASTQ file of UMI reads, matches the reads between the files based on read names, and
+    |adds optional tags to the BAM file to encode the UMIs for later use.
+  """,
+  group = ClpGroups.SamOrBam)
+class AnnotateUmis(
+  @arg(flag="i", doc="The input SAM or BAM file.")             val input: PathToBam,
+  @arg(flag="f", doc="Input FASTQ file with UMI reads.")       val fastq: PathToFastq,
+  @arg(flag="o", doc="Output BAM file to write.")              val output: PathToBam,
+  @arg(flag="t", doc="The BAM attribute to store UMIs in.")    val attribute: String = "BC",
+  @arg(          doc="If set, fail on the first missing UMI.") val failFast: Boolean = false
+) extends FgBioTool with LazyLogging {
+
+  private var missingUmis: Long = 0
+
+  /** Updates the count of missing UMI records, and throws an exception if fail-fast is true. */
+  private def logMissingUmi(readName: String) = {
+    missingUmis += 1
+    if (failFast) fail("Record '" + readName + "' in BAM file not found in FASTQ file.")
+  }
+
+  /** Main method that does the work of reading input files, matching up reads and writing the output file. */
+  override def execute(): Unit = {
+    Io.assertReadable(Seq(input, fastq))
+    Io.assertCanWriteFile(output)
+
+    // Read in the fastq file
+    logger.info("Reading in UMIs from FASTQ.")
+    val fqIn = new FastqReader(fastq.toFile)
+    val nameToUmi = fqIn.view.map(f => (f.getReadHeader.takeWhile(_ != ' '), f.getReadString)).toMap
+    fqIn.close()
+
+    // Loop through the BAM file an annotate it
+    logger.info("Reading input BAM and annotating output BAM.")
+    val in  = SamReaderFactory.make().open(input.toFile)
+    val out = new SAMFileWriterFactory().makeWriter(in.getFileHeader, true, output.toFile, null)
+    val progress = new ProgressLogger(logger)
```

Comment:
Shame that the `=`s are not aligned.

![olb0hx5](https://cloud.githubusercontent.com/assets/846274/13517264/c24fb710-e18f-11e5-88f6-334bff9481f0.gif)